### ISSUE TITLE
`Extent` parameter for ImageItem?

### ISSUE THREAD

**Issue by ntjess** [pyqtgraph/pyqtgraph#1543]
Maybe I missed it in the documentation, but is there a way to add an `extent` to an `ImageItem` (a la `pyplot.imshow(...extent=[xmin, xmax, ymin, ymax]?)` I am plotting spatial data and it would be great if my axes could line up with the true spatial values. Currently, I can add an offset to my image but I cannot show fractional axis steps per pixel.

Use case: Each pixel is 0.05mm from the next. I would like the x- and y-axes to represent mm, but I am not sure how to do this with the existing `ImageItem` support.

Thanks!

**Comment by ixjlyons** [comment]
[setRect](https://pyqtgraph.readthedocs.io/en/latest/graphicsItems/imageitem.html#pyqtgraph.ImageItem.setRect) should do what you want. It could be nice to add some flexibility in the API so you don't have to construct a `QRect(F)` yourself, and maybe add it to the options available via [setImage](https://pyqtgraph.readthedocs.io/en/latest/graphicsItems/imageitem.html#pyqtgraph.ImageItem.setImage).

```python
import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtCore

m_per_pixel = 0.05e-3
img_size = 600

app = pg.mkQApp()

win = pg.plot()
win.setAspectLocked(True)

data = np.random.normal(size=(img_size, img_size))
img = pg.ImageItem(data)
img.setRect(QtCore.QRectF(0, 0, img_size*m_per_pixel, img_size*m_per_pixel))

win.setLabel("bottom", text="x", units="m")
win.setLabel("left", text="y", units="m")

win.addItem(img)
app.exec_()
```

**Comment by ntjess** [comment]
worked like a charm, thanks!

**Comment by j9ac9k** [comment]
@ixjlyons what do you think the call to `setImage` should look like?

```python
img.setImage(data, extent=[1.5, 1.5])  # for x and y, in some fractional value?
```

**Comment by ntjess** [comment]
FWIW, copied from `matplotlib.imshow`:

```
extentfloats (left, right, bottom, top), optional
The bounding box in data coordinates that the image will fill. The image is stretched individually along x and y to fill the box.

The default extent is determined by the following conditions. Pixels have unit size in data coordinates. Their centers are on integer coordinates, and their center coordinates range from 0 to columns-1 horizontally and from 0 to rows-1 vertically.

Note that the direction of the vertical axis and thus the default values for top and bottom depend on origin:

For origin == 'upper' the default is (-0.5, numcols-0.5, numrows-0.5, -0.5).
For origin == 'lower' the default is (-0.5, numcols-0.5, -0.5, numrows-0.5).
See the origin and extent in imshow tutorial for examples and a more detailed description.
```

**Comment by j9ac9k** [comment]
usually I try and emulate the Qt API, but in absence of that, the matplotlib API is a fantastic API to emulate to fall back on.

Unrelated to this issue, I wanted to chat more w/ you about your application;... would you be up for hopping on our slack channel?

**Comment by ntjess** [comment]
Not at all! Do you have a link for me? Apologies if it is publicly available and I missed it.

**Comment by j9ac9k** [comment]
https://join.slack.com/t/pyqtgraph/shared_invite/zt-mi68iriu-HWwHStyC_f7fCItWUFO19w  Haven't publicly posted it, it's not exactly a private group.

**Comment by ixjlyons** [comment]
> @ixjlyons what do you think the call to `setImage` should look like?

Yeah, probably not a great candidate for the good first issue label because there's an API decision to be made. I was originally just thinking of making `setRect` take `*args` to accept either a `QRect[F]` or separate x/y/w/h for conveience, then adding a `rect` kwarg to `setImage` that could either be a `QRect[F]` or a tuple. Maybe instead `setImage` could take `xlim`/`ylim` tuples and there could be a different supporting method for that. Or just a single `scale` argument since I'd imagine that's the most common use case. Or just copy matplotlib :)

**Comment by NilsNemitz** [comment]
Including any alternative to QRectF would also make it obvious that this is supposed to be a user-accessible function. When I looked over the code and docs, I did not realize that `setRect()` was supposed to do exactly what I was looking for at the time :)
It would be nice to see this added!

I like the idea of supporting
`extent=(w,h)`   and   `extent=(x,y,w,h)`   as well as    `setExtent( (w,h) )` and `setExtent( (x,y,w,h) )`
in addition to the current QRectF.

Less important, but probably convenient:
`scale=scale` or `scale=(w_scale, h_scale)`

Regarding the setter functions:
How do we expect these to work when the image is updated?
My impression is that once I call `ImageItem.setScale(5)` on a persistent object, I think I would expect that a new displayed image is also magnified.
I am not sure what I would expect after calling e.g. `setExtent(0,0,100,100)`. Do all future images map into the assigned rectangle? The current behaviourseems to be to discard the setting from `setRect()` as soon as `setImage()` is called again.
It might be helpful to make this setting persistent across images. I think this is an additional argument for using "extent" as a new term and leaving `setRect()` with the current, non-persistent behavior.

**Comment by ixjlyons** [comment]
> Including any alternative to QRectF would also make it obvious that this is supposed to be a user-accessible function.

Great point, that's a nice benefit I hadn't thought about.

It occurred to me that there's [`QGraphicsItem.setScale`](https://doc.qt.io/qt-5/qgraphicsitem.html#setScale) and [`QGraphicsItem.setPos`](https://doc.qt.io/qt-5/qgraphicsitem.html#setPos), and they seem to work fine, except with `pxMode`. Maybe we just need to solve the issue of documenting parent methods (#1237).

**Comment by NilsNemitz** [comment]
I think this was documented with #1726, and the solutions should now be easier to find.
Should this be closed?
